MericCelik/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,291 @@
+     * the target is in the tree. If these conditions are not hold, it prints suitable
+     * warning message to the standart output.
+     * Complexities: O(1) when warnings occurs, otherwise depend on findsAncestors method.
+     * @param target : node whose ancestors are found.
+     * @return nothing.
+     * */
+
+    public void printAncestors(BinaryTreeNode<T> target){
+        if(target==null){
+            System.out.println("There is no target node!");
+            return;
+        }
+        if(this.getRoot()==null){
+            System.out.println("The tree is empty!");
+            return;
+        }
+
+        if(target.equals(this.root)){
+            System.out.println("Target node is the root of the tree, it has no parent!");
+            return;
+        }
+
+        ArrayList<BinaryTreeNode<T>> ancestors = findAncestors(target);
+        int ancestor_count = ancestors.size();
+        if(ancestor_count==0 || !this.getRoot().equals(ancestors.get(ancestor_count-1))){
+            System.out.println("This node is not in the binary tree!");
+            return;
+        }
+
+        for(int i=0; i<ancestor_count;i++)
+            System.out.print(ancestors.get(i).getKey() + " ");
+        System.out.println();
+    }
+
+    /**Calls the findAncestors method for two nodes and finds their lowest common ancestor by comparing
+     * resulting arrays. If one of them is in other's resulting array, this mean it is the ancestor of the other.
+     * Else if the array sizes are equal, it means they are in the same level. It traverse arrays simultaneously
+     * until it finds first common item. This is the lowest common ancestor.
+     * Else if the array sizes are not equal, it discards the deeper node's extra ancestors. It does it by incrementing
+     * the index value for the array traverse.
+     * The order of the nodes are not important.
+     * Complexities: O(1) when warnings occurs, otherwise depend on findsAncestors method.
+     * @param node1 : first node to consider
+     * @param node2 : other node to consider
+     * @return lowest common ancestor node
+     * */
+
+    public BinaryTreeNode<T> lowestCommonAncestor(BinaryTreeNode<T> node1, BinaryTreeNode<T> node2){
+        if(node1==null&&node2==null){
+            System.out.println("There are no first and second node!");
+            return null;
+        }
+        if(node1==null){
+            System.out.println("There is no first node!");
+            return null;
+        }
+        if(node2==null){
+            System.out.println("There is no second node!");
+            return null;
+        }
+        if(this.getRoot()==null){
+            System.out.println("The tree is empty!");
+            return null;
+        }
+
+        if(node1.equals(this.getRoot())&&node2.equals(this.getRoot())){
+            System.out.println("Both of them are root!");
+            return null;
+        }
+
+        ArrayList<BinaryTreeNode<T>> ancestors1 = findAncestors(node1);
```

Comment:
In the problem description it was mentioned that for problem 2.2 extra memory should not be used. Can you think of a different algorithm that would not use the list from 2.1? Perhaps going from one node upwards towards the root and checking if the other node is in that subtree?

Reply:
I am sorry, apparently, I misunderstood even though I asked on Slack. :sweat_smile: I thought that storing ancestors as a node's attribute was prohibited. I imagined it like this:
```
public class BinaryTreeNode<T> {
    private T key;
    private BinaryTreeNode<T> rightchild;
    private BinaryTreeNode<T> leftchild;
    private BinaryTreeNode<T> parent;
    private ArrayList<BinaryTreeNode<T>> ancestors;
}
```
When I think about it now, returning the ancestors using the function is the same as this. My mistake! :laughing: I tried to implement a solution similar to the one we discussed. I hope it is not wrong!